NIFI-3688 Extended Groovy Nifi Processor

dlukyanov opened the pull request:
Extended Groovy Nifi Processor

some samples
```groovy
//get attribute
def a = flowFile.AttributeName

//set attribute
flowFile.AttributeName = 'new value'

//get content input stream
flowFile.read()

//get content reader
flowFile.read().newReader("UTF-8")

//write through writer
flowFile.write("UTF-8"){ writer -> writer.append("Hello world!") }
flowFile.write("UTF-8", (java.lang.CharSequence) "Hello world!" )
flowFile.write("UTF-8", groovy.lang.Writable )

//write stream
flowFile.write{ outStream -> outStream << someInputStream }

//read rows from the database
//CTL.mydb must be declared as a processor property and linked to dbcp controller
def name = "World"
def rows = CTL.mydb.rows("select $name as dbname from dual")
assert rows.size() == 1
assert rows[0].DBNAME == 'World'

//write content to database blob
import groovy.sql.Sql
def theid = 123
def stream = flowFile.read()
CTL.mydb.execute("update mytable set myblob = ${ Sql.BLOB(stream) } where id=$theid")

```

### For all changes:
- [YES] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?
 in last one only

- [YES] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [YES] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [YES?] Is your initial contribution a single, squashed commit?
 
### For code changes:
- [YES] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [NO] Have you written or updated unit tests to verify your changes?
 no changes done to existing code. but no tests to groovyx processor.
- [YES] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [YES?] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [YES?] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [NO] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?
 not applicable?
### For documentation related changes:
- [YES] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

alopresto commented on nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-nar/pom.xml:
```diff
@@ -0,0 +1,44 @@
+    <artifactId>nifi-groovyx-nar</artifactId>
+    <packaging>nar</packaging>
+    <properties>
+        <maven.javadoc.skip>true</maven.javadoc.skip>
```
I don't think this is compatible with our coding style guide. Is there a reason Javadoc needs to be skipped?

dlukyanov replied:
no reason. i'll fix it.

mattyb149 replied:
I think for the NAR POM this is fine, there should be no Javadoc generated for the NAR itself. This is a pretty standard pattern for NARs, I'm guessing you got it from the scripting NAR's POM :)

joewitt commented on nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-nar/src/main/resources/META-INF/NOTICE:
```diff
@@ -0,0 +1,123 @@
+
+The following binary components are provided under the MIT License.  See project link for details.
+
+  (MIT License) EWS Java API
```
You don't need to reference CategoryA deps in the NOTICE which are already covered in the LICENSE.  This MIT dep and its copyright appears in the LICENSE already it seems so you can remove this section.

dlukyanov replied:
This just copy-paste from another processor. Could you point to the right one license header?

joewitt commented on nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-processors/src/main/java/org/apache/nifi/processors/groovyx/GroovyMethods.java:
```diff
@@ -0,0 +1,266 @@
+import org.apache.nifi.processor.Relationship;
+import org.apache.nifi.processor.FlowFileFilter;
+import org.apache.nifi.processor.FlowFileFilter.FlowFileFilterResult;
+//import org.apache.nifi.processor.ProcessSession;
```
please remove

joewitt commented:
The commit which removed the properties/skip javadoc/src does need to be in the nar bundle itself.  We dont want src/javadocs for the nar bundle.  Also, the processor needs a name that follows the pattern of VerbSubject instead of GroovyX.

We will definitely need unit tests for this.  There are references to Oracle things in there or at least in the screenshot.  The LICENSE and NOTICE appears to have a lot of copy/paste from other nars.  Are all those dependencies really there?

dlukyanov commented:
@joewitt , about naming: CallGroovy would be fine?
Actually it's another implementation of groovy script. 
And the Script processor is just a Script - and not CallScript.
I wanted to point that it's extended groovy script.. What do you think about the name?

dlukyanov commented:
Can somebody help me?
Avro tests fails all the time and I did not touch it.
Failed tests: 
  TestAvroRecordReader.testLogicalTypes:109 expected:<2017-04-0[4]> but was:<2017-04-0[5]>

alopresto commented:
@dlukyanov That is not your fault -- @mattyb149 is fixing it in [NIFI-3718](https://issues.apache.org/jira/browse/NIFI-3718).

mattyb149 commented:
I recommend ExecuteGroovyScript as the name, it seems to support the same use case as ExecuteScript but with Groovy-specific bells and whistles. I haven't had a time to review in depth, but will soon.

dlukyanov commented:
@mattyb149, I renamed the processor, created test cases for all mentioned in pull request description.
The case [test_sql_01_select.groovy](https://github.com/dlukyanov/nifi/blob/master/nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-processors/src/test/resources/groovy/test_sql_01_select.groovy) dedicated to a new way of controller services usage from groovy script.

mattyb149 commented on nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-processors/pom.xml:
```diff
@@ -0,0 +1,74 @@
+        <dependency>
+            <groupId>org.codehaus.groovy</groupId>
+            <artifactId>groovy-all</artifactId>
+            <version>2.4.7</version>
```
I believe 2.4.11 has been released, it is not required that you upgrade it here, just mentioning it :)

mattyb149 commented on nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-processors/src/main/java/org/apache/nifi/processors/groovyx/ExecuteGroovyScript.java:
```diff
@@ -0,0 +1,461 @@
+import org.apache.nifi.components.ValidationContext;
+
+@EventDriven
+@Tags({"script", "groovy", "groovyx", "extended"})
```
I'm not sure that users will search "extended" looking for this processor, perhaps this tag is not needed?

mattyb149 commented on nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-processors/src/main/java/org/apache/nifi/processors/groovyx/ExecuteGroovyScript.java:
```diff
@@ -0,0 +1,461 @@
+@CapabilityDescription(
+        "Experimental Extended Groovy script processor. The script is responsible for " + "handling the incoming flow file (transfer to SUCCESS or remove, e.g.) as well as any flow files created by "
+                + "the script. If the handling is incomplete or incorrect, the session will be rolled back.")
+@SeeAlso({})
```
This is not needed, recommend removal

mattyb149 commented on nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-processors/src/main/java/org/apache/nifi/processors/groovyx/flow/GroovySessionFile.java:
```diff
@@ -0,0 +1,284 @@
+
+    protected GroovySessionFile(ProcessSessionWrap session, FlowFile f) {
+        super(session, f);
+        setMetaClass(null); //set defult metaclass
```
Minor typo here (default vs defult)

mattyb149 commented:
This NAR won't be added to the NiFi distribution unless you add an entry for it in nifi-assembly/pom.xml, and set its version in the root level pom.xml (see those files for many examples).

mattyb149 commented:
Also there are still 22 checkstyle violations (mostly lines too long), you can see which files are affected by running "mvn clean install -Pcontrib-check" from your nifi-groovyx-bundle directory.

dlukyanov commented:
seems I fixed all checkstyle issues) but appveyor failed.. what should i do with it?

mattyb149 commented:
Appveyor always fails, no worries. I will take a look at your latest

dlukyanov commented:
@mattyb149 , @joewitt , @alopresto ,

I know that 1.2.0. released 
This pull request was not ready (

I need to know your opinion about this pull request - to continue or not.

In any case I fixed documentation and nar bundle content.
You can try all the cases from pull request description.

I built groovyx nar bundle with nifi 1.2.0
https://github.com/dlukyanov/nifi/releases/tag/nifi-groovyx-nar-1.2.0

mattyb149 commented on nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-nar/pom.xml:
```diff
@@ -0,0 +1,44 @@
+    <parent>
+        <groupId>org.apache.nifi</groupId>
+        <artifactId>nifi-groovyx-bundle</artifactId>
+        <version>1.2.0-SNAPSHOT</version>
```
Now that NiFi 1.2.0 has been released, the version of the current master branch is 1.3.0-SNAPSHOT. Do you mind rebasing your PR against master, and updating the version of this NAR to 1.3.0-SNAPSHOT wherever necessary? Please and thanks!

dlukyanov replied:
Great! I'll do it this weekend.

dlukyanov commented:
I've made last several commits in StandardFlowSerializerTest.java

trying to detect the problem java.nio.channels.OverlappingFileLockException in org.apache.nifi.controller.scheduling.TestProcessorLifecycle

https://issues.apache.org/jira/browse/NIFI-3853?focusedCommentId=16015655&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16015655

and succeed with build. 
normally should revert this change in my fork...

mattyb149 commented:
Don't worry about test failures in other parts of the project, they can be handled with other Jira cases and fixes. When I review your PR I build and test the relevant parts, and ensure a successful full build (whether all the tests pass or not).  Feel free to revert those commits and rebase the whole thing into one, will make it easier to review/merge, please and thanks!

dlukyanov commented:
In this case i believe it's ready for review.
I've reverted last commits and rebased all commits into one
Tried to build locally - and it's fine.

mattyb149 commented:
Sorry for the delay, I've been meaning to get back to this review. Unfortunately in the meantime the master branch has moved to version 1.4.0-SNAPSHOT, can you change your references from 1.3.0-SNAPSHOT to 1.4.0-SNAPSHOT and rebase against the latest master? Please and thank you!

mattyb149 commented on nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-processors/src/main/java/org/apache/nifi/processors/groovyx/ExecuteGroovyScript.java:
```diff
@@ -0,0 +1,468 @@
+            .build();
+
+    public static String[] VALID_BOOLEANS = {"true", "false"};
+    public static final PropertyDescriptor REQUIRE_FLOW = new PropertyDescriptor.Builder().name("Requires flow file")
```
I see the benefit of saving the user the trouble of the following boilerplate:

```
def flowFile = session.get()
if(!flowFile) return
```

However there are other methods on ProcessSession that might be helpful to the user, for example getting a batch of flow files (if available): 

```
def flowFiles = session.get(100)
if(!flowFiles) return
flowFiles.each { flowFile ->
  // do something
}
```
or applying a filter (I see GroovyProcessSessionWrap has a handy get(Closure) override but it is unused). For the small amount of boilerplate code, I'm not sure such a property is necessary. Thoughts?

dlukyanov replied:
with `require flow file = false` you can do any session.get(...) in code
i agree it's not a big win to use `require flow file = true`. 
i thought to add more values like `batch` and use `batch_size` custom property.

a short entry about `require flow file` and `fail strategy`:
we were migrating an 10years old j2ee project to nifi.  100k+ producers and 100k consumers of the files with different formats (xml,csv,json,pdf,tiff,...). you can imagine a huge amount of specific code, libraries, ... to cover all specific things we used groovy script. and in 95% we used just `session.get(0)` (require flow file)

and in 80%  `rollback` strategy  on exception is ok (btw it's a standard behavior of existing execute script processor). without try-catch the code just simpler. in case when using `rollback` you can do any custom try-catch in your code. and we used custom try-catch in 1% of cases.

transfer to `failure` in code usually becomes a nightmare in script especially if you want to transfer the original file - and in our case it was always original. 

so, both features are not removing customization ability. 

about the `require flowfile` - i agree that it could be removed.

but about `fail strategy` - i believe it's very useful. maybe the items in select should be renamed and more clarified... but i'd suggest to keep it.

mattyb149 commented on nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-processors/src/main/java/org/apache/nifi/processors/groovyx/ExecuteGroovyScript.java:
```diff
@@ -0,0 +1,468 @@
+            .required(true).expressionLanguageSupported(false).allowableValues(VALID_BOOLEANS).defaultValue("false").build();
+
+    public static String[] VALID_FAIL_STRATEGY = {"rollback", "transfer to failure"};
+    public static final PropertyDescriptor FAIL_STRATEGY = new PropertyDescriptor.Builder()
```
This also seems to try to automate "normal" handling of flow files, but removes a little flexibility. Inside the script, the user has the ability to roll back (via a thrown exception) or fail when desired. If the idea is to allow the flow files to be transferred to failure in the event of an exception in the script itself (rather than one caused by bad input, for example), then it might be confusing for the user to know why the flow file was transferred to failure (i.e. is the data bad or is the script bad).  Maybe I'm just misunderstanding the intent, could you clarify for me?

mattyb149 commented on nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-processors/src/main/java/org/apache/nifi/processors/groovyx/ExecuteGroovyScript.java:
```diff
@@ -0,0 +1,468 @@
+import org.apache.nifi.components.ValidationResult;
+import org.apache.nifi.components.ValidationContext;
+
+@EventDriven
```
You might want an InputRequirement annotation here, just for clarity. The processor will default to INPUT_ALLOWED which I think is what you want, but it would probably help here, especially in the context of your REQUIRE_FLOW property below

mattyb149 commented on nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-processors/src/main/java/org/apache/nifi/processors/groovyx/GroovyMethods.java:
```diff
@@ -0,0 +1,85 @@
+            }
+
+            /** to support: REL_SUCCESS << sessionFile */
+            private Relationship leftShift(Relationship r, SessionFile f) {
```
These are great!  Should we support right-shift too?

`sessionFile >> REL_SUCCESS`

If having both is too confusing, then this one will work just fine!

dlukyanov replied:
we can do both. at the beginning i did rightShift (more logical when coding). then i looked at http://docs.groovy-lang.org/latest/html/groovy-jdk/java/io/OutputStream.html and it implements only leftShift, so i decided to use leftShift. but i like rightShift also)

dlukyanov replied:
btw. leftshift allows syntax: 
`REL_SUCCESS << a << b << c`

dlukyanov commented on nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-processors/src/main/java/org/apache/nifi/processors/groovyx/ExecuteGroovyScript.java:
```diff
@@ -0,0 +1,468 @@
+        + "handling the incoming flow file (transfer to SUCCESS or remove, e.g.) as well as any flow files created by "
+        + "the script. If the handling is incomplete or incorrect, the session will be rolled back.")
+@Restricted("Provides operator the ability to execute arbitrary code assuming all permissions that NiFi has.")
+@SeeAlso({})
```
ok. it's just a copy-paste from somewhere)

dlukyanov commented:
@mattyb149, I fixed everything except:

- Failure strategy
 this property does not limit anything in error handling. it provides default algorithms for non-handled exceptions. the user still can do try-catch-transfer in code. i beleive i've added enough description of this property. ready for discussion about this point.
- rightShift
 i really like it, but the main reason why i did not do it - that in similar cases in groovy only `leftShift` used.
 -- http://docs.groovy-lang.org/latest/html/groovy-jdk/java/io/OutputStream.html
 -- http://docs.groovy-lang.org/latest/html/groovy-jdk/java/io/Writer.html
 -- http://docs.groovy-lang.org/latest/html/groovy-jdk/java/util/Collection.html

dlukyanov commented:
@mattyb149, do you have any comments to my last commit?

mattyb149 commented on nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-nar/pom.xml:
```diff
@@ -0,0 +1,44 @@
+    <parent>
+        <groupId>org.apache.nifi</groupId>
+        <artifactId>nifi-groovyx-bundle</artifactId>
+        <version>1.4.0-SNAPSHOT</version>
```
Sorry I keep losing track of this PR :(  In the meantime we have released 1.4.0, do you mind changing these references to 1.5.0-SNAPSHOT and rebasing against the latest master? Please and thanks!

dlukyanov commented:
mattyb149, rebased against the latest master 1.5.0-SNAPSHOT

mattyb149 commented on nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-processors/src/main/resources/docs/org.apache.nifi.processors.groovyx.ExecuteGroovyScript/additionalDetails.html:
```diff
@@ -0,0 +1,202 @@
+<p>
+CTL.* objects accessible if corresponding processor property defined.<br/>
+<b>Example:</b> if you defined property <code>`CTL.cache`</code> to DistributedMapCacheClientService, then you can access it from code <code>CTL.cache</code><br/>
+If CTL property references to Database connection pool, then corresponding CTL entry will contain groovy.sql.Sql object connected to database with autocommit=false.</br>
```
How is it guaranteed that autocommit will be false? If setAutoCommit is called somewhere (either by your code -- which I couldn't find a reference to -- or by groovy's Sql class), it can cause problems in a couple of scenarios, one being Oracle if the DBA has disallowed changing auto-commit, and another is Hive (since HiveConnectionPool extends DBCPService, it should be available via CTL right?)

dlukyanov replied:
Here I call `setAutoCommit(false)`: [ExecuteGroovyScript.java#L312](https://github.com/dlukyanov/nifi/blob/master/nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-processors/src/main/java/org/apache/nifi/processors/groovyx/ExecuteGroovyScript.java#L312) 

with hive that's interesting question. according to code [HiveConnection.setAutoCommit()](https://github.com/apache/hive/blob/master/jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java#L1338) will log warning. but the [HiveConnection.commit()](https://github.com/apache/hive/blob/master/jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java#L873) will not work in any case. What do you think? Allow `groovy.sql.Sql` to be created only for `DBCPService` ?

mattyb149 replied:
The version of setAutoCommit() you link to is their master branch. The version that ships with NiFi's Hive NAR is [here](https://github.com/apache/hive/blob/release-1.2.1/jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java#L1197). Strangely, they're handling it backwards from how they do in master.
Just to be safe, it might be better to not explicitly call setAutoCommit(false) and maybe instead allow the script writer to call it on CTL or the appropriate object?

This is your processor and contribution, so you can choose whether to support JDBC connections that don't behave with the calls you're making. Hive is just one example, but there might be others that don't implement certain methods well or at all. It is totally fine to just say "those aren't supported", I'm just bringing it up in case you wanted to address such things. Let me know either way and I'll continue the review/merge, thanks!

mattyb149 commented on nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-processors/src/main/resources/docs/org.apache.nifi.processors.groovyx.ExecuteGroovyScript/additionalDetails.html:
```diff
@@ -0,0 +1,202 @@
+</p>
+<p><i>set new attribute value</i></p>
+<pre>
+  flowFile.ATTRIBUTE_NAME = ATTRIBUTE_VALUE 
```
These features are awesome!

mattyb149 commented on nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-processors/src/main/resources/docs/org.apache.nifi.processors.groovyx.ExecuteGroovyScript/additionalDetails.html:
```diff
@@ -0,0 +1,202 @@
+<h2>CTL map</h2>
+<p>
+CTL.* objects accessible if corresponding processor property defined.<br/>
+<b>Example:</b> if you defined property <code>`CTL.cache`</code> to DistributedMapCacheClientService, then you can access it from code <code>CTL.cache</code><br/>
```
Can you explain (at least to me here) more about linking with CTL? Does the name of the service have to be "cache" in this case? Or does there need to be a user-defined property called "CTL.cache" added to the service? If the latter, what if the service itself uses the user-defined properties? Seems like there might be conflict in processing?

dlukyanov replied:
Yes, this adds kind of limit. All custom properties with name beginning with `CTL.` automatically supposed to be linked to controller services. It means you can't define custom `CTL.*` property with plain string value. After `CTL.`there could be any chars. On the level of groovy script the additional predefined variable binded : `CTL` - a hashmap with substring after `CTL.` as a key and linked controller service as a value. If you defined `CTL.aaa` and linked it to distributed cache client service, then on level of groovy it'slike `def CTL=[:]` and `CTL['aaa']=LINK_TO_THE_SERVICE`. So, in groovy you can access this service  `CTL.aaa.someServiceMethod(...)`

The only conflict expected when you'll try to define custom script variable with name `CTL`. Otherwise-no conflicts.

mattyb149 commented on nifi-nar-bundles/pom.xml:
```diff
@@ -94,7 +95,7 @@
         <plugins>
             <!-- Makes buildRevision and buildBranch available to the NAR Plugin so they can be populated in the MANIFEST -->
-            <plugin>
+            <!--plugin>
```
If this is not needed it should be removed

dlukyanov replied:
fixed (uncommented)

mattyb149 commented on nifi-nar-bundles/nifi-groovyx-bundle/nifi-groovyx-processors/src/main/java/org/apache/nifi/processors/groovyx/ExecuteGroovyScript.java:
```diff
@@ -0,0 +1,453 @@
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.controller.ControllerService;
+import org.apache.nifi.dbcp.DBCPService;
+import org.apache.nifi.flowfile.FlowFile;
```
Checkstyle (via the -Pcontrib-check Maven profile) says this is unused, so it should be removed

dlukyanov replied:
fixed

dlukyanov commented:
mattyb149, i just refactored the the `CTL` properties: 
- 'CTL' provides direct access to controller services without any additional logic
- a new map named `SQL` provides fast to transactional DBCP service

mattyb149 commented:
+1 LGTM, thanks for sticking with this, it's a great addition!  I made one final commit for some CheckStyle and typo stuff (and added displayName() to the properties), then rebased into one commit with you as the author.  Thanks again for this feature! Merging to master

mattyb149 commented:
Also I ran a full build with unit tests, and tried various scripts and features to exercise the processor.
